keps/sig-api-machinery/35-storage-version-hash.md

```diff
@@ -0,0 +1,163 @@
+Hence, we use the hash value instead of the plain text to avoid clients misusing
+the information. The hash function needs to ensure hash value differs if the
+storage versions are different.
+
+[APIResource]:https://github.com/kubernetes/kubernetes/blob/f22334f14d92565ec3ff9d4ff2b995eae9af622a/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L881-L905
+
+```golang
+type APIResource struct {
+        // The hash value of the storage version, the version this resource is
+        // converted to when written to the data store. Value must be treated 
+        // as opaque by clients. Only equality comparison on the value is valid.
+        // This field is of alpha quality and might be removed in the next
+        // release.
+        // +optional
+        StorageVersionHash string `json:"storageVersionHash,omitempty"`
+        // These are the existing fields.
+        Name string
+        SingularName string
+        Namespaced bool
+        Group string
+        Version string
+        Kind string
+        Verbs Verbs
+        ShortNames []string
+        Categories []string
+}
+```
+
+### Implementation details
+
+The hash function needs to ensure hash value differs if the storage versions are
+different. SHA-256 is good enough for our purpose.
+
+For different categories of resources,
+
+* For built-in resources, the kube-apiserver will set the `StorageVersionHash`
+when it bootstraps.
+
+* For custom resources, the storage version is already exposed through the [CRD
+spec][]. The apiextension apiserver will hash the value and sets the
+`StorageVersionHash` when registers the discovery doc for the CRD.
+
+* For aggregated resources, if the aggregated apiserver is implemented using the
+generic apiserver library, the `StorageVersionHash` will be set in the same way
+as the kube-apiserver. Otherwise, the aggregated apiserver is responsible
+to come up with a proper `StorageVersionHash`.
+
+* For sub-resources, the `StorageVersionHash` field will be left empty. No
+sub-resource persists in the data store and thus does not require storage
+version migration.
+
+[CRD spec]:https://github.com/kubernetes/kubernetes/blob/7d8554643e2e05fda714f30fc71f34ce05514b68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L167
+
+## Graduation Criteria
+
+The discovery API is read-only and the `StorageVersionHash` field is only
+intended to be used by the storage version migrator, the graduation story is
+simple.
+
+We will document the field as of alpha quality in 1.14. If we don't find any
+problem with the field, we will promote it to beta in 1.15 and GA in 1.16.
+Otherwise we just remove the field while keeping a tombstone for the protobuf
+tag.
+
+The above is a simplified version of Kubernetes API change [guideline][].
+
+[guideline]:https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version
+
+## Risks and mitigation
+
+### HA masters
```

Comment:
> Does this field actually provide value to you before the HA master bits are settled?

I think so, in the following ways:
* Auto-triggered migration is useful for non-HA clusters
* Manually triggered migration after an HA upgrade is complete could determine if the storage version for a particular resource had changed since the last run and skip unnecessary migrations